In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sqlalchemy import create_engine

In [2]:
api = pd.read_csv("C:\\Users\\JSLV3\\Documents\\datos_descargados.csv")
api.head()


fecha_terminacion_proyecto           fecha_de_corte  \
0    2018-12-31T00:00:00.000  2024-04-04T00:00:00.000   
1    2018-10-16T00:00:00.000  2024-04-04T00:00:00.000   
2    2018-10-08T00:00:00.000  2024-04-04T00:00:00.000   
3    2019-06-01T00:00:00.000  2024-04-04T00:00:00.000   
4    2019-02-12T00:00:00.000  2024-04-04T00:00:00.000   

   c_digo_divipola_departamento departamento c_digo_divipola_municipio  \
0                            18      CAQUETA       18029, 18205, 18610   
1                            15       BOYACA                     15693   
2                            17       CALDAS                     17050   
3                            17       CALDAS                     17446   
4                             5    ANTIOQUIA                      5250   

                                           municipio  \
0  Albania(CAQ), Curillo(CAQ), San Jose De Fragua...   
1                         Santa Rosa De Viterbo(BOY)   
2                                      Aranzazu(CAL)   
3                                     Marulanda(CAL)   
4                                      El Bagre(ANT)   

                                           indicador  \
0       Nueva Población Beneficiada Acueducto Urbano   
1       Nueva Población Beneficiada Acueducto Urbano   
2  Nueva Población Beneficiada Alcantarillado Urbano   
3        Nueva Población Beneficiada Acueducto Rural   
4       Nueva Población Beneficiada Acueducto Urbano   

                                     nombre_proyecto origen  aporte_nacion  \
0  CONSTRUCCION SISTEMA DE ACUEDUCTO REGIONAL PAR...    PDA    12110573518   
1  CONSTRUCCION RED DE CONDUCCION, REDES DE DISTR...    PDA      557345219   
2  OPTIMIZACION SISTEMAS DE ACUEDUCTO Y ALCANTARI...    PDA      491820256   
3  MEJORAMIENTO SISTEMA DE ACUEDUCTO EN EL CORREG...    PDA      286128295   
4  CONSTRUCCION DEL PLAN MAESTRO DE ACUEDUCTO URB...    PDA              0   

   contrapartida estado_seguimiento  
0      953340488     En Liquidación  
1     2236932650     En Liquidación  
2      664819896          Terminado  
3      540512429          Terminado  
4     4773790842          Liquidado

In [3]:
api.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 268 entries, 0 to 267
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   fecha_terminacion_proyecto    268 non-null    object
 1   fecha_de_corte                268 non-null    object
 2   c_digo_divipola_departamento  268 non-null    int64 
 3   departamento                  268 non-null    object
 4   c_digo_divipola_municipio     268 non-null    object
 5   municipio                     268 non-null    object
 6   indicador                     268 non-null    object
 7   nombre_proyecto               268 non-null    object
 8   origen                        268 non-null    object
 9   aporte_nacion                 268 non-null    int64 
 10  contrapartida                 268 non-null    int64 
 11  estado_seguimiento            268 non-null    object
dtypes: int64(3), object(9)
memory usage: 25.3+ KB


In [10]:
api.describe(include='all')


fecha_terminacion_proyecto       fecha_de_corte  \
count                             268                  268   
unique                            NaN                  NaN   
top                               NaN                  NaN   
freq                              NaN                  NaN   
mean    2020-12-24 21:18:48.358209024  2024-04-04 00:00:00   
min               2017-03-08 00:00:00  2024-04-04 00:00:00   
25%               2019-08-11 00:00:00  2024-04-04 00:00:00   
50%               2020-11-08 00:00:00  2024-04-04 00:00:00   
75%               2022-01-05 06:00:00  2024-04-04 00:00:00   
max               2024-02-27 00:00:00  2024-04-04 00:00:00   
std                               NaN                  NaN   

        c_digo_divipola_departamento departamento c_digo_divipola_municipio  \
count                     268.000000          268                       268   
unique                           NaN           29                       194   
top                              NaN       Nariño                     44430   
freq                             NaN           30                         6   
mean                       43.298507          NaN                       NaN   
min                         5.000000          NaN                       NaN   
25%                        20.000000          NaN                       NaN   
50%                        44.000000          NaN                       NaN   
75%                        63.000000          NaN                       NaN   
max                        95.000000          NaN                       NaN   
std                        24.237443          NaN                       NaN   

          municipio                                          indicador  \
count           268                                                268   
unique          193                                                  4   
top     Maicao(La )  Nueva Población Beneficiada Alcantarillado Urbano   
freq              6                                                 98   
mean            NaN                                                NaN   
min             NaN                                                NaN   
25%             NaN                                                NaN   
50%             NaN                                                NaN   
75%             NaN                                                NaN   
max             NaN                                                NaN   
std             NaN                                                NaN   

                                          nombre_proyecto origen  \
count                                                 268    268   
unique                                                248     11   
top     CONSTRUCCIÓN DE CONEXIONES INTRADOMICILIARIAS ...    PDA   
freq                                                    4    141   
mean                                                  NaN    NaN   
min                                                   NaN    NaN   
25%                                                   NaN    NaN   
50%                                                   NaN    NaN   
75%                                                   NaN    NaN   
max                                                   NaN    NaN   
std                                                   NaN    NaN   

        aporte_nacion  contrapartida estado_seguimiento  
count    2.680000e+02   2.680000e+02                268  
unique            NaN            NaN                  4  
top               NaN            NaN          Terminado  
freq              NaN            NaN                102  
mean     2.759527e+09   2.829967e+09                NaN  
min      0.000000e+00   0.000000e+00                NaN  
25%      0.000000e+00   0.000000e+00                NaN  
50%      3.800891e+08   1.277241e+09                NaN  
75%      3.302808e+09   3.914618e+09                NaN  
max      6.152706e+10   3.11268

•	Fecha de terminación del proyecto: Las fechas varían desde el 8 de marzo de 2017 hasta el 27 de febrero de 2024, con una fecha media de terminación alrededor del 24 de diciembre de 2020.

•	Fecha de corte: Todas las observaciones tienen la misma fecha de corte, el 4 de abril de 2024.

•	Código DIVIPOLA del departamento: Hay 29 categorías únicas, con Nariño teniendo la mayor frecuencia (30 veces).

•	Municipio: Hay 193 municipios únicos mencionados, con 'Maicao (La Guajira)' apareciendo más frecuentemente (6 veces).

•	Indicador: Hay 4 indicadores diferentes de beneficios proporcionados por los proyectos. El más común es "Nueva Población Beneficiada Alcantarillado Urbano" (98 veces).

•	Nombre del proyecto: Hay 248 nombres de proyectos únicos, con el nombre más común siendo sobre construcción de conexiones intradomiciliarias, apareciendo 4 veces.

•	Origen: Hay 11 orígenes diferentes de financiamiento, con PDA siendo el más frecuente (141 veces).

•	Aporte de la nación y contrapartida: Estos campos son numéricos con valores que varían ampliamente, reflejando la diversidad de escala y tipo de financiamiento de los proyectos.

•	Estado de seguimiento: Los proyectos tienen uno de cuatro estados de seguimiento, siendo "Terminado" el más común (102 veces).


Conversión de Fechas

In [5]:
api['fecha_terminacion_proyecto'] = pd.to_datetime(api['fecha_terminacion_proyecto'])
api['fecha_de_corte'] = pd.to_datetime(api['fecha_de_corte'])

Las columnas fecha_terminacion_proyecto y fecha_de_corte estaban originalmente en formato de texto (string), lo cual no es óptimo para realizar operaciones relacionadas con fechas y tiempos, así que, convertimos estas columnas al tipo datetime.

Normalización de Textos en Columnas Categóricas

In [12]:
str_cols = api.select_dtypes(include=['object']).columns

api[str_cols] = api[str_cols].apply(lambda x: x.str.lower().str.strip())

Las discrepancias en el formato de texto, como las diferencias entre mayúsculas y minúsculas o la presencia de espacios en blanco adicionales, pueden crear categorías artificiales que distorsionan el análisis. Por ejemplo, "CALDAS", "Caldas" y " caldas " se tratarían como tres categorías separadas cuando, en realidad, representan la misma entidad.
Para abordar estos problemas y garantizar la consistencia en el análisis, realizamos la siguiente normalización en todas las columnas de texto:

1.	Convertimos todas las cadenas de texto a minúsculas, asegurando que la capitalización no cree duplicados artificiales.

2.	Eliminamos los espacios en blanco al principio y al final de las cadenas, evitando discrepancias debidas a errores de formato o tipografía.


Verificación de Valores Nulos

In [9]:
api.isnull().sum()

fecha_terminacion_proyecto      0
fecha_de_corte                  0
c_digo_divipola_departamento    0
departamento                    0
c_digo_divipola_municipio       0
municipio                       0
indicador                       0
nombre_proyecto                 0
origen                          0
aporte_nacion                   0
contrapartida                   0
estado_seguimiento              0
dtype: int64

Los resultados confirman que no hay ausencia de datos en ninguna parte del conjunto de datos.

 Filas Duplicadas

In [8]:
api.duplicated().sum()

0

El resultado de esta operación es cero, lo que indica que no hay filas duplicadas en nuestro conjunto de datos.

In [11]:
api['num_municipios'] = api['c_digo_divipola_municipio'].apply(lambda x: len(x.split(',')))


La nueva columna num_municipios fue creada para cuantificar el alcance de cada proyecto en términos del número de municipios afectados. En el dataset original, la columna c_digo_divipola_municipio contenía listas de códigos de municipios, indicando que algunos proyectos tienen influencia sobre múltiples localidades. Sin embargo, esta información estaba condensada en una sola cadena de texto, lo cual dificulta realizar análisis estadísticos o comparativos sobre ella.

Agrupación de Proyectos por Región Geográfica

In [15]:
region_mapping = {
    'CHOCO': 'Región Pacífica',
    'CAUCA': 'Región Pacífica',
    'NARIÑO': 'Región Pacífica',
    'VALLE DEL CAUCA': 'Región Pacífica',
    'ARAUCA': 'Región Orinoquía',
    'CASANARE': 'Región Orinoquía',
    'GUAINÍA': 'Región Orinoquía',
    'GUAVIARE': 'Región Orinoquía',
    'META': 'Región Orinoquía',
    'VICHADA': 'Región Orinoquía',
    'AMAZONAS': 'Región Amazonía',
    'CAQUETA': 'Región Amazonía',
    'PUTUMAYO': 'Región Amazonía',
    'VAUPES': 'Región Amazonía',
    'ANTIOQUIA': 'Región Andina',
    'BOYACA': 'Región Andina',
    'CALDAS': 'Región Andina',
    'CUNDINAMARCA': 'Región Andina',
    'HUILA': 'Región Andina',
    'NORTE DE SANTANDER': 'Región Andina',
    'QUINDIO': 'Región Andina',
    'RISARALDA': 'Región Andina',
    'SANTANDER': 'Región Andina',
    'TOLIMA': 'Región Andina',
    'SAN ANDRES Y PROVIDENCIA': 'Región Insular',
    'ATLANTICO': 'Región Caribe',
    'BOLIVAR': 'Región Caribe',
    'CESAR': 'Región Caribe',
    'CORDOBA': 'Región Caribe',
    'LA GUAJIRA': 'Región Caribe',
    'MAGDALENA': 'Región Caribe',
    'SUCRE': 'Región Caribe'
}

# Corrección de nombres en el dataset
api['departamento'] = api['departamento'].replace({
    'SAN ANDRES': 'SAN ANDRES Y PROVIDENCIA',
    'N DE SANTANDER': 'NORTE DE SANTANDER'
})

# Aplicación del mapeo para crear la columna de región
api['región'] = api['departamento'].str.upper().map(region_mapping)

# Manejo de posibles valores no mapeados
api['región'] = api['región'].fillna('Región Desconocida')

# Mostrar el mapeo aplicado
api[['departamento', 'región']]



departamento            región
0        caqueta   Región Amazonía
1         boyaca     Región Andina
2         caldas     Región Andina
3         caldas     Región Andina
4      antioquia     Región Andina
..           ...               ...
263    santander     Región Andina
264       boyaca     Región Andina
265     casanare  Región Orinoquía
266    risaralda     Región Andina
267   la guajira     Región Caribe

[268 rows x 2 columns]

La columna creada "región", representa una clasificación estratégica de los departamentos en regiones geográficas más amplias. El propósito de esta agrupación es simplificar el análisis de los datos y permitir una interpretación más clara de las tendencias y patrones a un nivel regional, en lugar de hacerlo por departamentos individuales, que son más numerosos y podrían conducir a un análisis más fragmentado y complejo.